### PR TITLE
CPT: Infinitely render CPTs using ReactVirtualized

### PR DESCRIPTION
**Pull request by aduth**
Supersedes #5007

This pull request seeks to implement infinite scroll for the custom post types listing screen.

__Implementation notes:__

This takes a unique approach to loading more items. Rather than triggering a request based on the scroll approaching the next page, it leverages [ReactVirtualized's `onRowsRendered` callback](https://github.com/bvaughn/react-virtualized/blob/master/docs/VirtualScroll.md) to determine which pages should be requested for the query dependent on which rows are currently rendered. The reason for this change is to accommodate persistence of posts queries state, which is not currently persisted. Once persisted, we need a mechanism to responsibly refresh all of the persisted data. In the context of having persisted post entries, traditional "needs more items so fetch" logic does not work well, and naively re-requesting all persisted posts could incur huge network costs if tens of pages had been loaded during the last session.

This pull request also upgrades ReactVirtualized to ~~7.5.0~~ 7.9.1 to make use of the `WindowScroller` HOC ([see release notes](https://github.com/bvaughn/react-virtualized/blob/master/CHANGELOG.md#750)).

__Testing instructions:__

Verify that all posts are revealed on the custom post types list screen, requested 20 at a time infinitely until the last known page. When changing filters, note that the same behavior applies to the new filter.

1. Navigate to [My Sites](http://calypso.localhost:3000/stats/insights)
2. Select a site where at least one custom post type is enabled
3. Choose the custom post type navigation link in the sidebar
4. Note that a loading placeholder or, if persisted state exists, the past results are immediately shown
5. Note that posts are revealed ten at a time
6. Note that as you scroll, more posts are loaded until all pages have been exhausted

/cc @timmyc 

Test live: https://calypso.live/?branch=update/cpt-infinite-list-2

### CODE REVIEW

File: client/my-sites/post-type-list/index.jsx [hunk 1]
```diff
@@ -11,50 +18,151 @@ import QueryPosts from 'components/data/query-posts';
...
-					</li>
+/**
+ * Constants
+ */
+const POST_ROW_HEIGHT = 86;
+const DEFAULT_POSTS_PER_PAGE = 20;
+const LOAD_OFFSET = 10;
+
+class PostTypeList extends Component {
+	static propTypes = {
```
Comment (timmyc) [inline client/my-sites/post-type-list/index.jsx]: fancy 🎉

**Comment by timmyc** [comment]
Okay took this one for a spin, and as expected I was able to find a nice edge case where things went down the road to _strange_ town.

Although I did encounter a few other oddities, this is one flow that I could reproduce consistently.  But here is the video:

mystery-draft

The flow:
- View CPT listing page
- Click Add in the sidebar
- Add a title, some content, wait for autosave to trigger
- Click back button
- If your last view was the draft list just wait, if it was the published filter click drafts
- The draft testimonial will not show up, and the loading indicator keeps spinning
- In the network tab you can verify the API has returned the payload with the correct data

If you toggle back to "Published" and then to "Drafts" again, the new draft will show up - and the loading indicator goes away.

I did test publishing new back-dated posts and they show up in their expected order in the list though which is pretty sweet!!

**Comment by aduth** [comment]
Thanks @timmyc ! I think the issue you discovered is essentially identical to the one described at https://github.com/Automattic/wp-calypso/pull/5647#issuecomment-224045776 . This should be resolved in bab5428f66d8099a49e27343113627e5b3cb0aaa . Mind taking it for another spin?

**Comment by aduth** [comment]
We might also be able to do something as simple as `this.props.posts !== prevProps.posts` (since posts state isn't mutated), though this might cause some unnecessary forced updates.

File: client/state/posts/selectors.js [hunk 2]
```diff
@@ -115,7 +115,12 @@ export function getSitePostsLastPageForQuery( state, siteId, query ) {
 		return null;
 	}
 
-	return state.posts.queries[ siteId ].getNumberOfPages( query );
+	const pages = state.posts.queries[ siteId ].getNumberOfPages( query );
```
Comment (timmyc) [inline client/state/posts/selectors.js]: Is there any chance `state.posts.queries[ siteId ]` could ever be null or undefined?

**Comment by timmyc** [comment]
That did fix the issue, nice!  So while testing the same flow, I've been hitting an exception when navigating `back` to the post-types-list:

testimonials_ _testingtimmy2_wordpress_com_ _wordpress_com

the line it is failing on

testimonials_ _testingtimmy2_wordpress_com_ _wordpress_com

Though just tested the same flow on master and I'm getting that same exception in about 3 out of 4 tests.

As such I'll just create a new issue for that bug :)

This is looking good to me - just one new minor question above, otherwise 🚢

File: client/state/posts/selectors.js (hunk 2 above)
Comment (aduth) [inline client/state/posts/selectors.js]: >Is there any chance `state.posts.queries[ siteId ]` could ever be null or undefined?

Yes, and it's obscured by the diff on GitHub, but a few lines above we're checking for this.

**Comment by aduth** [comment]
Rebased, and in the course of some final testing uncovered an oddity worth a more detailed look, reported in #6120. In the context of these changes, it manifested itself in a loader always being shown as a contributor trying to load drafts, because this component isn't able to reconcile the fact that no posts were received, but there's supposedly 2 pages of data to expect. On the posts screen, we'd be shown the "No drafts found" warning if no posts were included in the response. If it turns out to be a REST API issue, I'd rather address it there, so I'll not let it hold back the merge of this one.

**Comment by aduth** [comment]
Not sure if you experienced this yourself @timmyc, but I felt the scrolling was not very smooth here. I dug a bit deeper and it may be at least partly attributable to `react-virtualized`. I opened an issue at bvaughn/react-virtualized#281 .

File: client/my-sites/post-type-list/index.jsx [hunk 3]
```diff
@@ -11,52 +18,171 @@ import QueryPosts from 'components/data/query-posts';
...
+				{ query && this.state.requestedPages.map( ( page ) => (
+					<QueryPosts
+						key={ `query-${ page }` }
+						siteId={ siteId }
+						query={ { ...query, page } } />
 				) ) }
-			</ul>
-		</div>
-	);
+				{ isEmpty && (
```
Comment (timmyc) [inline client/my-sites/post-type-list/index.jsx]: While working on the `TermSelector` I discovered the `noRowsRenderer` option for `VirtualScroll`.  Not sure if you want to leverage that here to show `PostTypeListEmptyContent` and/or loading state.

**Comment by aduth** [comment]
@timmyc Thanks for pointing me to `noRowsRenderer`. I looked to see if this would make sense for the posts listing, but I decided against using it. For the `<EmptyContent />` rendering, I'd rather not put the burden on `react-virtualized` for rendering that, since it'd have to do its magic to calculate its own height to make sure it's displayed in its entirety. Seems simpler to just render it separate. For loading states, we'll want to consider case where we've loaded persisted data, in which case items will exist and we want a loading placeholder to be shown at the top of the list (`items` length is not zero in this case, yet we want the initial placeholder).

That said, I found that my change in 29c0951 caused a regression in the editor post selector where it had relied on `lastPage` being `false` when no results were found. I don't think this was the right way to treat it, so I updated it in 50a7569 to account for the new behavior. In this instance, `noRowsRenderer` was a perfect fit.

**Comment by aduth** [comment]
Upgraded once more to react-virtualized 7.9.1, specifically for [the 7.8.3 release](https://github.com/bvaughn/react-virtualized/releases/tag/7.8.3) which includes some minor performance improvements from yours truly (bvaughn/react-virtualized#283).

I doubled-back on my stance to not use `noRowsRenderer` in b1f42d0 because of another edge case I found where navigating from a list where a single result had been loaded to a new page would not trigger the request for the new query. This should hopefully be a bit clearer to understand than the previous condition in `componentDidUpdate`.

**Pull request merged**